lib/tag-suggestions/index.jsx

```diff
@@ -0,0 +1,85 @@
+import React, { Component, Fragment } from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import appState from '../flux/app-state';
+import { tracks } from '../analytics';
+
+const { search } = appState.actionCreators;
+const { recordEvent } = tracks;
+
+export class TagSuggestions extends Component {
+  static displayName = 'TagSuggestions';
+
+  static propTypes = {
+    query: PropTypes.string.isRequired,
+    filteredTags: PropTypes.array.isRequired,
+  };
+
+  render() {
+    const { filteredTags } = this.props;
+
+    return (
+      <Fragment>
+        {filteredTags.length > 0 && (
+          <div className="tag-suggestions">
+            <div className="note-list-header">Tags</div>
+            <ul className="tag-suggestions-list">
+              {filteredTags.map(tag => (
+                <li
+                  key={tag.id}
+                  id={tag.id}
+                  className="tag-suggestion-row"
+                  onClick={() => this.props.onSearch(`tag:${tag.id}`)}
+                >
+                  <div
+                    className="tag-suggestion"
+                    title={decodeURIComponent(tag.id)}
+                  >
+                    tag:{decodeURIComponent(tag.id)}
+                  </div>
+                </li>
+              ))}
+            </ul>
+          </div>
+        )}
+      </Fragment>
+    );
+  }
+}
+
+const mapStateToProps = ({ appState: state }, ownProps) => ({
+  filteredTags: state.tags
+    .filter(function(tag) {
+      // todo split on spaces to support additive query args
+
+      // prefix tag ID with "tag:"; this allows us to match if the user typed the prefix
+      // n.b. doing it in this direction instead of stripping off any "tag:" prefix allows support
+      // of tags that contain the string "tag:" ¯\_(ツ)_/¯
+      var testID = 'tag:' + tag.data.name;
+
+      // exception: if the user typed "tag:" or some subset thereof, don't return all tags
+      if (['t', 'ta', 'tag', 'tag:'].includes(ownProps.query)) {
+        testID = tag.data.name;
+      }
+
+      return (
+        testID.search(new RegExp('(tag:)?' + ownProps.query, 'i')) !== -1 &&
+        // discard exact matches -- if the user has already typed or clicked
+        // the full tag name, don't suggest it
+        testID !== ownProps.query
+      );
+    })
+    .slice(0, 5),
+});
```

Comment:
I'd like to try and make some updates to help you here, but there's one thing we should be careful of here and that's how this will cause the tag suggestions list to re-render on every state change even if it doesn't need to. the reason is that every time we `.filter` or `.slice` we create a new array in JavaScript and Redux is going to use _shallow equality_ to determine if things change.

that means that even though we return the same values, it will be treated as a brand new list of tags here.
the solution to prevent this is to return the same _reference-equal_ value when nothing changes. this is typically done in a number of ways: through memoization or through moving the filtered tags into Redux state.

the easier change here is memoization and that would be a cache based on the list of tags and the filter.

```js
const filterTags = ( tags, query ) => tags.filter( tag => {
	…
} ).slice( 0, 5 );

let lastTags = null;
let lastQuery = null;
let lastMatches = [];
const getMatchingTags = ( tags, query ) => {
	if ( lastTags === tags && lastQuery === query ) {
		return lastMatches;
	}

	lastTags = tags;
	lastQuery = query;
	lastMatches = filterTags( tags, query );
	return lastMatches;
}

const mapStateToProps = ( { appState: { filter, tags } } ) => ( {
	filteredTags: getMatchingTags( tags, filter ),
} );
```